include/caliper/ReportInfrastructure.hpp

```diff
@@ -0,0 +1,141 @@
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/// \file  Report.cpp
+/// \brief Generates text reports from Caliper snapshots on flush() events 
+
+#ifndef CALIPER_COMMON_REPORTINFRASTRUCTURE_HPP
+#define CALIPER_COMMON_REPORTINFRASTRUCTURE_HPP
+
+#include "caliper/Caliper.h"
+#include "caliper/SnapshotRecord.h"
+
+#include "caliper/reader/Format.h"
+#include "caliper/reader/RecordSelector.h"
+#include "caliper/reader/Table.h"
+
+#include "caliper/common/Log.h"
+#include "caliper/common/Node.h"
+#include "caliper/common/RuntimeConfig.h"
+#include "caliper/common/SnapshotTextFormatter.h"
+
+#include <iostream>
+#include <fstream>
+#include <sstream>
+
+using namespace cali;
+
+namespace cali
+{
+namespace reporting
+{
+    
+    class FileBufferStream : public std::basic_streambuf<char>{
+      FILE* fp;
+      public:
+      FileBufferStream(FILE* m_fp) : fp(m_fp) {}
+      virtual int_type overflow(int_type ch){
+        fputc((char)ch, fp);
+        return 0;
+      }
+    };
+
+    class Reporter {
+        using output_stream_type = std::ostream*;
+        Caliper m_cali;
+        output_stream_type m_output_stream;
+        Table          m_table_writer;
+        RecordSelector m_selector;
+
+        std::vector<Entry> make_entrylist(Caliper* c, const SnapshotRecord* snapshot) {
+            std::vector<Entry> list;
+
+            if (!snapshot)
+                return list;
+
+            SnapshotRecord::Data  data(snapshot->data());
+            SnapshotRecord::Sizes size(snapshot->size());
+
+            for (size_t n = 0; n < size.n_nodes; ++n)
+                list.push_back(Entry(data.node_entries[n]));
+            for (size_t n = 0; n < size.n_immediate; ++n)
+                list.push_back(Entry(c->get_attribute(data.immediate_attr[n]), data.immediate_data[n]));
+
+            return list;
+        }
+
+        void process_snapshot(Caliper* c, const SnapshotRecord* snapshot) {
+            SnapshotProcessFn fn(m_table_writer);
+            m_selector(*c, make_entrylist(c, snapshot), fn);
+        }
+
+        void flush(Caliper* c, const SnapshotRecord* flush_info) {
+            m_table_writer.flush(*c, *m_output_stream);
+        }
+        public:
+        void report(){
+          m_table_writer.flush(m_cali, *m_output_stream);
+        }
+        Reporter(output_stream_type out, const std::string& attributes, const std::string& sort, const std::string& filter, Caliper c = Caliper::instance()) :
+              m_cali(c),
```

Comment:
Don't copy the Caliper instance object please! Just get a new one when needed. (In fact, I should make the copy constructor private.)